server/src/main/scala/org/http4s/server/middleware/Throttle.scala

```diff
@@ -0,0 +1,90 @@
+package org.http4s.server.middleware
+
+import org.http4s.{Http, Response, Status}
+import cats.data.Kleisli
+import cats.effect.{Concurrent, Timer}
+import cats.effect.concurrent.Ref
+import scala.concurrent.duration._
+import cats.implicits._
+import fs2.Stream
+
+sealed trait TokenAvailability
+case object TokenAvailable extends TokenAvailability
+case object TokenUnavailable extends TokenAvailability
+
+/**
+  * A token bucket for use with the [[Throttle]] middleware.  Consumers can take tokens which will be refilled over time.
+  * Implementations are required to provide their own refill mechanism.
+  *
+  * Possible implementations include a remote TokenBucket service to coordinate between different application instances.
+  */
+trait TokenBucket[F[_]] {
+  def takeToken: F[TokenAvailability]
+}
+
+object TokenBucket {
+
+  /**
+    * Creates an in-memory [[TokenBucket]].
+    *
+    * @param capacity the number of tokens the bucket can hold and starts with.
+    * @param refillEvery the frequency with which to add another token if there is capacity spare.
+    * @return A singleton stream containing the [[TokenBucket]].
+    */
+  def local[F[_]](capacity: Int, refillEvery: FiniteDuration)(
+      implicit F: Concurrent[F],
+      timer: Timer[F]): Stream[F, TokenBucket[F]] =
+    Stream.eval(Ref[F].of(capacity)).flatMap { counter =>
+      val refill = Stream
```

Comment:
I've implemented this previously by storing the last time a token was requested.  Then you can interpolate out how much has been refilled from last time before deciding whether or not to grant a token, and not need a background process.  This approach results in fractional tokens.

I think we'd also then get away with an `F[TokenBucket[F]]` instead of a `Stream[F, TokenBucket[F]]`.

Reply:
> I think we'd also then get away with an F[TokenBucket[F]] instead of a Stream[F, TokenBucket[F]].

we can do this regardless btw. Returning a Resource[TokenBucket]` by `compile.drain.start` and `cancel` as the shutdown action.

You mentioned the refill approach a couple times but never with enough detail to understand _how_ you did it, so this is interesting :)

Reply:
Just out of curiosity, it wouldn't work with a TokenBucket that had the task waiting rather than rejected, right? (if more recent requests don't arrive, old tasks are stuck)

Reply:
I think my rejection message included a "retry-after" like value.  I don't think I maintained a queue.  In either case, we can compute the time until full and try again.  It does get messy if we're returning the same retry-after to multiple requestors and get slammed.